File: Packages/NotebarCore/Sources/NotebarCore/Models/DiagnosticsEnvironment.swift
import Foundation

/// The environment half of an Export Diagnostics file (spec §6.4b) — app
/// version, OS version, display geometry, and the database's own
/// `DatabaseDiagnostics` snapshot. Deliberately holds nothing else: every
/// field is a fact about the machine or the store, never a note or task's
/// content, so `renderedText` is safe to hand to anyone regardless of what
/// is actually in the database. The app target gathers these values (they
/// need `Bundle`, `ProcessInfo`, `NSScreen` — none of it AppKit-only except
/// the screen geometry strings, which are already reduced to plain text by
/// the time they reach here) and appends the unified-log excerpt itself,
/// which this type has no part in.
public struct DiagnosticsEnvironment: Equatable, Sendable {
    public var appVersion: String
    public var buildNumber: String
    public var osVersion: String
    /// One line per connected display, already formatted as plain text
    /// (e.g. `"1728x1117 @ (0, 0), scale 2.0"`) — computed by the caller so
    /// this module never needs to know what an `NSScreen` is.
    public var displayGeometry: [String]
    public var database: DatabaseDiagnostics

    public init(
        appVersion: String,
        buildNumber: String,
        osVersion: String,
        displayGeometry: [String],
        database: DatabaseDiagnostics
    ) {
        self.appVersion = appVersion
        self.buildNumber = buildNumber
        self.osVersion = osVersion
        self.displayGeometry = displayGeometry
        self.database = database
    }

    /// Plain-text rendering for the diagnostics file. No note or task
    /// content ever passes through this type's stored properties, so
    /// nothing here can leak it — see `DiagnosticsTests` for the invariant
    /// this depends on holding at the call site too (the caller must never
    /// pass note content into any of these fields, which none of them have
    /// room for in the first place).
    public var renderedText: String {
        var lines = [
            "App version: \(appVersion) (\(buildNumber))",
            "macOS version: \(osVersion)",
        ]
        if displayGeometry.isEmpty {
            lines.append("Displays: none reported")
        } else {
            lines.append("Displays:")
            lines.append(contentsOf: displayGeometry.map { "  - \($0)" })
        }
        lines.append("Database path: \(database.path ?? "(in-memory — on-disk store could not be opened)")")
        if let size = database.sizeOnDisk {
            lines.append("Database size: \(size) bytes")
        } else {
            lines.append("Database size: unknown")
        }
        if database.appliedMigrations.isEmpty {
            lines.append("Migrations applied: none")
        } else {
            lines.append("Migrations applied: \(database.appliedMigrations.joined(separator: ", "))")
        }
        return lines.joined(separator: "\n")
    }
}

File: windows/Notebar.Core/Models/DiagnosticsEnvironment.cs
using System.Text;

namespace Notebar.Core.Models;

/// <summary>The environment half of an Export Diagnostics file.</summary>
/// <remarks>
/// Deliberately holds nothing else: every field is a fact about the machine or
/// the store, never a note or task's content, so RenderedText is safe to hand to
/// anyone regardless of what is actually in the database. The app gathers these
/// values and appends its own log excerpt, which this type has no part in.
/// </remarks>
public sealed record DiagnosticsEnvironment(
    string AppVersion,
    string BuildNumber,
    string OsVersion,
    /// <summary>One line per connected display, already formatted as plain text
    /// (e.g. "1920x1080 @ (0, 0), scale 1.5") — computed by the caller so this
    /// type never needs to know what a monitor handle is.</summary>
    IReadOnlyList<string> DisplayGeometry,
    DatabaseDiagnostics Database)
{
    public string RenderedText
    {
        get
        {
            var sb = new StringBuilder();
            sb.AppendLine($"App version: {AppVersion} ({BuildNumber})");
            sb.AppendLine($"Windows version: {OsVersion}");
            if (DisplayGeometry.Count == 0)
            {
                sb.AppendLine("Displays: none reported");
            }
            else
            {
                sb.AppendLine("Displays:");
                foreach (var line in DisplayGeometry) sb.AppendLine($"  - {line}");
            }
            sb.AppendLine($"Database path: {Database.Path ?? "(in-memory — on-disk store could not be opened)"}");
            sb.AppendLine(Database.SizeOnDisk is { } size
                ? $"Database size: {size} bytes"
                : "Database size: unknown");
            sb.AppendLine(Database.AppliedMigrations.Count == 0
                ? "Migrations applied: none"
                : $"Migrations applied: {string.Join(", ", Database.AppliedMigrations)}");
            return sb.ToString().TrimEnd();
        }
    }
}
